homework/index.js

```diff
@@ -29,23 +31,127 @@
+    addRow(table, 'Forks:', repo.forks, 'td');
+    addRow(table, 'Updated:', date.toLocaleDateString(), 'td');
+  }
+
+  function createOptions(repo, parent, index) {
+    createAndAppend('option', parent, {
+      text: repo.name,
+      value: index,
+    });
+  }
+  function selections(selection, data, parent) {
+    const ul = createAndAppend('ul', parent, {
+      class: 'ul-root',
+    });
+    const repo = data[selection];
+    renderRepoDetails(repo, ul);
+  }
+  function createContributorsSection(url, parent) {
+    fetchJSON(url).then(data => {
+      data.forEach(repo => {
+        const li = createAndAppend('li', parent);
+        createAndAppend('img', li, {
+          src: repo.avatar_url,
+          class: 'avatar',
+        });
+        createAndAppend('a', li, {
+          text: repo.login,
+          href: repo.html_url,
+          target: '_blank',
+          class: 'cont-name',
+        });
+        createAndAppend('p', li, {
+          text: repo.contributions,
+        });
+      });
+    });
   }
 
   function main(url) {
-    fetchJSON(url, (err, repos) => {
-      const root = document.getElementById('root');
-      if (err) {
+    const root = document.getElementById('root');
+    const header = createAndAppend('header', root, {
+      class: 'hyf-title',
+    });
+    createAndAppend('p', header, {
+      text: 'HYF Repositories',
+      class: 'main-title',
+    });
+    const mainSection = createAndAppend('main', root, {
+      class: 'main-container',
+    });
+    const repoContainerSection = createAndAppend('section', mainSection, {
+      class: 'repo-container',
+    });
+    const ul = createAndAppend('ul', repoContainerSection, {
+      class: 'ul-root',
+    });
+    const contributorsContainerSection = createAndAppend(
+      'section',
+      mainSection,
+      {
+        class: 'contributor-container',
+      },
+    );
+    const p = createAndAppend('p', contributorsContainerSection, {
+      class: 'contributors-title',
+    });
+    p.innerHTML = 'Contributions';
+
+    const ulCont = createAndAppend('ul', contributorsContainerSection, {
+      class: 'ul-cont-root',
+    });
+
+    const select = createAndAppend('select', header, {
+      class: 'selection-module',
+    });
+
+    fetchJSON(url)
+      .then(data => {
+        renderRepoDetails(data[17], ul);
+        const urlStart = data[17].contributors_url;
+        createContributorsSection(urlStart, ulCont);
+        data
+          .sort((a, b) => a.name.localeCompare(b.name))
+          .forEach((repo, index) => {
+            createOptions(repo, select, index);
+          });
+        select.addEventListener('change', () => {
+          repoContainerSection.innerHTML = '';
+          ulCont.innerHTML = '';
+          const selection = select.value;
+          selections(selection, data, repoContainerSection);
+          const contUrl = data[selection].contributors_url;
+          createContributorsSection(contUrl, ulCont);
+        });
```

Comment:
Basically you need to execute the same code to render a repository in two different places:

1. Once, at application start-up, using the repository at index zero.
2. Inside the change event handler, using the index of the selected repo.

When you detect a repeated code like this you should immediately consider writing a function to remove this duplication. Can try and write such a function?

Reply:
With the modified functions, your code could look like this:

```js
...
fetchJSON(url)
  .then(repos => {
    repos
      .sort((a, b) => a.name.localeCompare(b.name))
      .forEach((repo, index) => {
        createOption(repo, select, index);
      });
    changeSelection(repos[0], div, ulCont);

    select.addEventListener('change', () => {
      changeSelection(repos[select.value], div, ulCont);
    });
  })
  ...
```